### ISSUE TITLE
use The Update Framework (TUF)

### ISSUE THREAD

**Issue by zookoatleastauthoritycom** [zcash/zcash#170]
A couple of weeks ago, the day that Andrew left IEEE ("Oakland") and came to stay with Nathan, Andrew and Brian and I had a conversation about "forward-compatibility" and "future-proofing" features for ~Zcoin~ *Zcash* (see [1]). Out of that conversation, only two of the options seemed to me to be "shallow rabbit holes", where I could see the bottom of how many decisions and how much work we'd have to do in order to implement the thing: shipping with The Update Framework (TUF) http://theupdateframework.com/ (or any other comparable secure software updater), and having the ~Zcoin~ *Zcash* node freeze with a warning message if it detects that it is on the short side of a fork. This ticket is about the first one of those. https://github.com/Electric-Coin-Company/zerocashd/issues/171 is about the second one.

[1] [**Edit 2017-09-11:** Renamed to Zcash and added this footnote.] At the time this ticket was created, we called the project "Zcoin" as a tentative working name. This was prior to the Moneta project renaming to Zcoin project, which came out of an earlier project called Moneta, see: https://angel.co/moneta which links to moneta.cash which redirects to zcoin.tech.

**Comment by paragonie-scott** [comment]
> *security:
>     * dev signing keys are a juicier target
>     * larger attack surface

Counterpoint: According to [Verizon's Data Breach Incident Report for 2015](http://www.verizonenterprise.com/DBIR/2015/), most attacks conducted were the consequence of outdated software.

While theoretically the attack surface increases with every line of code (increase in space), you'd be significantly narrowing the gap for which a 1day exploit could be useful (decrease in time). I'd argue that this is a trade-off worth making, but (like everything else) one that should be taken with great care.

**Comment by indolering** [comment]
I've been over this a few times in the past.  As noted above, auto-updates amount to a backdoor but unpatched security flaws let attackers in through the frontdoor.  You can mitigate against backdoors through sandboxing and distributed trust protocols.  If I understand them correctly, TUF helps with the latter through what looks like a well thought out multi-party signature scheme.  

What I *don't* like about TUF is that (AFAIK) it doesn't bother with ensuring that signers haven't colluded to produce a custom backdoored update targeting a specific client.  This is fairly irritating since we've had blockchains for a decade now.

For end users, this is a no-brainer.  End users don't update their shit, ever, and it should be a felony for cloud providers to supply VMs without auto-updates turned on for security patches.  I don't know if TUF is the way to go, but you really should do this.  (It looks like you are working on Snap packages?) 

As to Nathan's points:

> another thing to explain to people about how/why we are different than Bitcoin

You can still download the raw binaries or build via Gitian.  FWIW, the last time I checked their multi-sig scheme was ad-hoc and manual.

> which features do we update to? "why are we forcing everyone to adopt feature X? Why are we evil?"

Controversial changes only kick in at specified dates and after % signaling.  Plenty of time to alter client config.

> legal controls, guarantees, warranties, etc.

Post update hashes to blockchain and have clients check before updating.  We can't comply with secret government orders to supply a backdoor in the form of an update if the client will refuse to accept it.

> more centralized decision making

More bikeshedding by people who don't pay for development.

> dev signing keys are a juicier target

They are already a pretty juicy target.

> larger attack surface

Agreed, it looks like there are OCaml and Haskell implementations though...

> engineering:
> more development time

Yeah, they don't really do GPG interop AFAIK.

> more code complexity, bugs, security flaws

Yup.

**Comment by paragonie-scott** [comment]
> What I don't like about TUF is that (AFAIK) it doesn't bother with ensuring that signers haven't colluded to produce a custom backdoored update targeting a specific client. This is fairly irritating since we've had blockchains for a decade now.

There's prior work here, by the way:

* https://paragonie.com/blog/2016/05/keyggdrasil-continuum-cryptography-powering-cms-airship
* https://wiki.mozilla.org/Security/Binary_Transparency

**Comment by indolering** [comment]
@paragonie-scott: Yeah, there should just be a PKI 101 CSE course so we can stop rehashing these same primitives over and over again.  The last "real" innovation was PoW solving sybil attacks.

Also, WTF Wordpress ... MD5 hashes?  You know they are still running OAuth 1.0?  I've gotta create a name-and-shame password hash website....

**Comment by paragonie-scott** [comment]
(To be completely fair, it's 8192 rounds of MD5 with a salt.)

@defuse wrote a post titled [The Triangle of Secure Code Delivery](https://defuse.ca/triangle-of-secure-code-delivery.htm) which is a good substitute for a 101 course for the purposes of this conversation.

**Comment by indolering** [comment]
@paragonie-scott Yeah, it was a great post!  I came up with a similar system while working on Namecoin :)

(Argon2 would bring the crack speed down from millions of guesses per second/GPU to hundreds)
